string.c

```diff
@@ -4605,71 +4605,68 @@ rb_str_aref(VALUE str, VALUE indx)
- *  the string.
- *
- *  Returns +nil+ if the regular expression does not match or the match string
- *  cannot be found.
- *
- *     a = "hello there"
- *
- *     a[1]                   #=> "e"
- *     a[2, 3]                #=> "llo"
- *     a[2..3]                #=> "ll"
- *
- *     a[-3, 2]               #=> "er"
- *     a[7..-2]               #=> "her"
- *     a[-4..-2]              #=> "her"
- *     a[-2..-4]              #=> ""
- *
- *     a[11, 0]               #=> ""
- *     a[11]                  #=> nil
- *     a[12, 0]               #=> nil
- *     a[12..-1]              #=> nil
- *
- *     a[/[aeiou](.)\1/]      #=> "ell"
- *     a[/[aeiou](.)\1/, 0]   #=> "ell"
- *     a[/[aeiou](.)\1/, 1]   #=> "l"
- *     a[/[aeiou](.)\1/, 2]   #=> nil
- *
- *     a[/(?<vowel>[aeiou])(?<non_vowel>[^aeiou])/, "non_vowel"] #=> "l"
- *     a[/(?<vowel>[aeiou])(?<non_vowel>[^aeiou])/, "vowel"]     #=> "e"
- *
- *     a["lo"]                #=> "lo"
- *     a["bye"]               #=> nil
+ *    string[index] -> new_string or nil
+ *    string[start, length] -> new_string or nil
+ *    string[range] -> new_string or nil
+ *    string[regexp] -> new_string or nil
+ *    string[regexp, capture] -> new_string or nil
+ *    string[substring] -> new_string or nil
+ *
+ *  When the single \Integer argument +index+ is given,
+ *  returns the 1-character substring found in +self+ at offset +index+:
+ *    'bar'[2] # => "r"
+ *  Counts backward from the end of +self+ if +index+ is negative:
+ *    'foo'[-3] # => "f"
+ *  Returns +nil+ if +index+ is out of range:
+ *    'foo'[3] # => nil
+ *    'foo'[-4] # => nil
+ *
+ *  When the two \Integer arguments  +start+ and +length+ are given,
+ *  returns the substring of the given +length+ found in +self+ at offset +start+:
+ *    'foo'[0, 2] # => "fo"
+ *    'foo'[0, 0] # => ""
+ *  Counts backward from the end of +self+ if +start+ is negative:
+ *    'foo'[-2, 2] # => "oo"
+ *  Special case: returns a new empty \String if +start+ is equal to the length of +self+:
+ *    'foo'[3, 2] # => ""
+ *  Returns +nil+ if +start+ is out of range:
+ *    'foo'[4, 2] # => nil
+ *    'foo'[-4, 2] # => nil
+ *  Returns the trailing substring of +self+ if +length+ is large:
+ *    'foo'[1, 50] # => "oo"
+ *  Returns +nil+ if +length+ is negative:
+ *    'foo'[0, -1] # => nil
+ *
+ *  When the single \Range argument +range+ is given,
+ *  derives +start+ and +length+ values from the given +range+,
+ *  and returns values as above:
+ *  - <tt>'foo'[0..1]</tt> is equivalent to <tt>'foo'[0, 2]</tt>.
+ *  - <tt>'foo'[0...1]</tt> is equivalent to <tt>'foo'[0, 1]</tt>.
+ *
+ *  When the single Regexp argument  +regexp+ is given,
+ *  returns the first matching substring found in +self+,
+ *  or +nil+ if none found:
+ *    'foo'[/o/] # => "o"
+ *    'foo'[/x/] # => nil
+ *
+ *  When arguments +regexp+ (a \Regexp)
+ *  and +capture+ (an \Integer capture group index or \String capture group name),
+ *  returns the \Matchdata as a \String
```

Comment:
I think this should be more specific.  If the capture is 0, it returns the entire MatchData as a string.  If the capture is a non-zero integer, it returns the numbered capture group, and if it is a string or symbol, the named capture group.

You may want to document the capture = 0 behavior in the previous section and move the first two example lines in this section to the section above.

Reply:
I'm not sure how this should be.  Do you want to make the change here?

Reply:
I think the examples that return the entire capture should be moved to the section above, along with appropriate explanation in the section above.  Alternatively, combine the section above and this into a single section. 

As currently written, the documentation is disjointed, because you have the first section talk about behavior with a single regexp (1 argument), and the second section about regexp+capture (2 arguments), but the first two examples in the second section are more similar to the behavior in the first section.

Something like:

```
 *  When the \Regexp argument +regexp+ is given,
 *  and the +capture+ argument is <tt>0</tt>,
 *  returns the first matching substring found in +self+,
 *  or +nil+ if none found:
 *    'foo'[/o/] # => "o"
 *    'foo'[/x/] # => nil
 *    s = 'hello there'
 *    s[/[aeiou](.)\1/] # => "ell"
 *    s[/[aeiou](.)\1/, 0] # => "ell"
 *
 *  If argument +capture+ is given and not <tt>0</tt>,
 *  it should be either an \Integer capture group index or a \String or \Symbol capture group name;
 *  the method call returns only the specified capture
 *  (see {Regexp Capturing}[Regexp.html#class-Regexp-label-Capturing]):
 *    s = 'hello there'
 *    s[/[aeiou](.)\1/, 1] # => "l"
 *    s[/(?<vowel>[aeiou])(?<non_vowel>[^aeiou])/, "non_vowel"] # => "l"
 *    s[/(?<vowel>[aeiou])(?<non_vowel>[^aeiou])/, :vowel] # => "e"

 *  If an invalid capture group index is given, +nil+ is returned.  If an invalid
 *  capture group name is given, +IndexError+ is raised.
```